File: 5_ispalindrome.py
def is_palindrome(s):
    # Если длина строки <= 1, строка - палиндром
    if len(s) <= 1:
        return True
    # Сверяем первый и последний символ
    if s[0] != s[-1]:
        return False
    # рекурсия строки со 2-го по предпоследний символ
    return is_palindrome(s[1:-1])

print(is_palindrome('rewqwer')) # True
print(is_palindrome('rewqwe')) # False

File: 5_ispalindrome.cpp
#include <iostream>
#include <string>

bool isPalindrome(const std::string& s, int start, int end) {
    // Если start >= end, все элементы проверены (длина <= 1)
    if (start >= end) {
        return true;
    }
    // Сверяем первый и последний символы
    if (s[start] != s[end]) {
        return false;
    }
    // рекурсия без первого и последнего символов
    return isPalindrome(s, start + 1, end - 1);
}

int main() {
    std::string test1 = "racecar";
    std::string test2 = "hello";

    std::cout << test1 << " палиндром? " << (isPalindrome(test1, 0, test1.length() - 1) ? "Да" : "Нет") << std::endl;
    std::cout << test2 << " палиндром? " << (isPalindrome(test2, 0, test2.length() - 1) ? "Да" : "Нет") << std::endl;

    return 0;
}
